tests/test_scheduler.py

```diff
@@ -0,0 +1,345 @@
+
+
+class TestSchedulerWithDownloaderAwareInMemory(BaseSchedulerInMemoryTester,
+                                               unittest.TestCase):
+    priority_queue_cls = 'scrapy.pqueues.DownloaderAwarePriorityQueue'
+
+    def test_logic(self):
+        downloader = self.mock_crawler.engine.downloader
+        for url, slot in _SLOTS:
+            request = Request(url)
+            downloader._set_slot_key(slot, request, None)
+            self.scheduler.enqueue_request(request)
+
+        slots = list()
+        requests = list()
+        while self.scheduler.has_pending_requests():
+            request = self.scheduler.next_request()
+            slot = downloader._get_slot_key(request, None)
+            slots.append(slot)
+            downloader.increment(slot)
+            requests.append(request)
+        self.assertEqual(len(slots), len(_SLOTS))
+
+        for request in requests:
+            slot = downloader._get_slot_key(request, None)
+            self.mock_crawler.engine.downloader.decrement(slot)
+
+        unique_slots = len(set(s for _, s in _SLOTS))
+        for i in range(0, len(_SLOTS), unique_slots):
+            part = slots[i:i + unique_slots]
+            self.assertEqual(len(part), len(set(part)))
+
+
+def _is_slots_unique(base_slots, result_slots):
+    unique_slots = len(set(s for _, s in base_slots))
+    for i in range(0, len(result_slots), unique_slots):
+        part = result_slots[i:i + unique_slots]
+        assert len(part) == len(set(part))
+
+
+class TestSchedulerWithDownloaderAwareOnDisk(BaseSchedulerOnDiskTester,
+                                             unittest.TestCase):
+    priority_queue_cls = 'scrapy.pqueues.DownloaderAwarePriorityQueue'
+
+    def test_logic(self):
+        downloader = self.mock_crawler.engine.downloader
+
+        for url, slot in _SLOTS:
+            request = Request(url)
+            downloader._set_slot_key(slot, request, None)
+            self.scheduler.enqueue_request(request)
+
+        self.close_scheduler()
+        self.create_scheduler()
+
+        slots = []
+        requests = []
+        downloader = self.mock_crawler.engine.downloader
+        while self.scheduler.has_pending_requests():
+            request = self.scheduler.next_request()
+            slot = downloader._get_slot_key(request, None)
+            slots.append(slot)
+            downloader.increment(slot)
+            requests.append(request)
+
+        self.assertEqual(len(slots), len(_SLOTS))
+
+        for request in requests:
+            slot = downloader._get_slot_key(request, None)
+            downloader.decrement(slot)
+
+        _is_slots_unique(_SLOTS, slots)
+        self.assertEqual(sum(len(s.active) for s in downloader.slots.values()), 0)
+
+
+class StartUrlsSpider(Spider):
+
+    def __init__(self, start_urls):
+        self.start_urls = start_urls
+
+    def parse(self, response):
+        pass
+
+
+class TestIntegrationWithDownloaderAwareOnDisk(TestCase):
```

Comment:
If I'm not mistaken, disk queues are not used in this test, because JOBDIR is not set.